File: backend/src/Services/SSEHubClient.php
<?php

declare(strict_types=1);

namespace Quantis\AIPortfolioAssistant\Services;

use Quantis\AIPortfolioAssistant\Contracts\StreamingClientInterface;
use Quantis\AIPortfolioAssistant\Exceptions\StreamingException;

/**
 * Server-Sent Events (SSE) client for streaming responses
 */
class SSEHubClient implements StreamingClientInterface
{
    private string $sessionId;
    private ?string $hubUrl;
    private bool $debug;
    private bool $connected = false;
    private bool $headersSet = false;

    public function __construct(string $sessionId, ?string $hubUrl = null, bool $debug = false)
    {
        $this->sessionId = $sessionId;
        $this->hubUrl = $hubUrl;
        $this->debug = $debug;
    }

    /**
     * Static factory method
     */
    public static function create(string $sessionId, ?string $hubUrl = null, bool $debug = false): self
    {
        return new self($sessionId, $hubUrl, $debug);
    }

    /**
     * Initialize SSE headers for direct streaming
     */
    public function initializeHeaders(): void
    {
        if ($this->headersSet) {
            return;
        }

        if (headers_sent()) {
            $this->log("Headers already sent, marking as initialized");
            $this->headersSet = true;
            $this->connected = true;
            return;
        }

        header('Content-Type: text/event-stream');
        header('Cache-Control: no-cache');
        header('Connection: keep-alive');
        header('X-Accel-Buffering: no');

        // Disable output buffering
        while (ob_get_level()) {
            ob_end_flush();
        }

        $this->headersSet = true;
        $this->connected = true;
        $this->log("SSE headers initialized for session: {$this->sessionId}");
    }

    /**
     * Mark headers as already initialized (when headers are set externally)
     */
    public function markHeadersInitialized(): void
    {
        $this->headersSet = true;
        $this->connected = true;
        $this->log("Headers marked as initialized for session: {$this->sessionId}");
    }

    /**
     * Send a progress message to the client
     */
    public function sendProgress(string $message): void
    {
        $this->sendEvent('progress', $message);
    }

    /**
     * Send a response (JSON data) to the client
     */
    public function sendResponse(array $data): void
    {
        $this->sendEvent('response', json_encode($data));
    }

    /**
     * Send an error to the client
     */
    public function sendError(string $message, int $code = 500): void
    {
        $this->sendEvent('error', json_encode([
            'error' => true,
            'message' => $message,
            'code' => $code
        ]));
    }

    /**
     * Send a chunk of streaming text
     */
    public function sendChunk(string $text): void
    {
        // DEBUG: Log EXACT data being sent to frontend (json_encode shows \n as \\n)
        error_log("📤 SENDING TO FRONTEND (len=" . strlen($text) . "): " . json_encode($text));
        $this->sendEvent('chunk', $text);
    }

    /**
     * Signal that streaming is complete
     */
    public function complete(): void
    {
        $this->sendEvent('complete', json_encode(['status' => 'done']));
        $this->connected = false;
    }

    /**
     * Send a custom event with arbitrary data
     */
    public function sendCustomEvent(string $eventName, array $data): void
    {
        $this->sendEvent($eventName, json_encode($data));
    }

    /**
     * Get the session ID
     */
    public function getSessionId(): string
    {
        return $this->sessionId;
    }

    /**
     * Check if the client is connected
     */
    public function isConnected(): bool
    {
        return $this->connected && connection_status() === CONNECTION_NORMAL;
    }

    /**
     * Send an SSE event
     */
    private function sendEvent(string $event, string $data): void
    {
        if ($this->hubUrl) {
            $this->sendToHub($event, $data);
        } else {
            $this->sendDirect($event, $data);
        }
    }

    /**
     * Send event directly via echo (for direct streaming)
     */
    private function sendDirect(string $event, string $data): void
    {
        if (!$this->headersSet) {
            $this->initializeHeaders();
        }

        echo "event: {$event}\n";

        // SSE spec: if data contains newlines, send each line as separate "data:" line
        $lines = explode("\n", $data);
        foreach ($lines as $line) {
            echo "data: {$line}\n";
        }
        echo "\n";  // End of event

        if (ob_get_level()) {
            ob_flush();
        }
        flush();

        $this->log("Direct SSE sent - event: {$event}");
    }

    /**
     * Send event to an SSE hub server
     */
    private function sendToHub(string $event, string $data): void
    {
        try {
            $payload = json_encode([
                'session_id' => $this->sessionId,
                'event' => $event,
                'data' => $data
            ]);

            $ch = curl_init($this->hubUrl);
            curl_setopt_array($ch, [
                CURLOPT_POST => true,
                CURLOPT_POSTFIELDS => $payload,
                CURLOPT_HTTPHEADER => [
                    'Content-Type: application/json',
                    'Content-Length: ' . strlen($payload)
                ],
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_TIMEOUT => 5
            ]);

            $result = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            curl_close($ch);

            if ($httpCode !== 200) {
                $this->log("Hub send failed with HTTP {$httpCode}");
            }
        } catch (\Throwable $e) {
            $this->log("Hub send error: " . $e->getMessage());
        }
    }

    /**
     * Log a debug message
     */
    private function log(string $message): void
    {
        if ($this->debug) {
            error_log("[SSEHubClient] {$message}");
        }
    }
}

File: backend_python/app/services/sse_hub_client.py
"""Port of Services/SSEHubClient.php.

Server-Sent Events (SSE) client for streaming responses.

The stream is obtained from the `stream` constructor kwarg or, when omitted,
from the `current_stream` ContextVar (set by `ChatController` in the request
thread before calling the assistant — thread-local because the whole request
runs in one thread).
"""
from __future__ import annotations

import contextvars

import httpx

from app.contracts.streaming_client import StreamingClientInterface
from app.support.logger import error_log
from app.support.phpjson import dumps
from app.support.sse import SseStream

current_stream: contextvars.ContextVar[SseStream | None] = contextvars.ContextVar('sse_current_stream', default=None)


class SSEHubClient(StreamingClientInterface):
    current_stream = current_stream

    def __init__(self, sessionId: str, hubUrl: str | None = None, debug: bool = False, stream: SseStream | None = None):
        self.sessionId = sessionId
        self.hubUrl = hubUrl
        self.debug = debug
        self.connected = False
        self.headersSet = False
        self._stream = stream

    @classmethod
    def create(cls, sessionId: str, hubUrl: str | None = None, debug: bool = False, stream: SseStream | None = None) -> 'SSEHubClient':
        """Static factory method."""
        return cls(sessionId, hubUrl, debug, stream=stream)

    def _get_stream(self) -> SseStream | None:
        return self._stream if self._stream is not None else current_stream.get()

    def markHeadersInitialized(self) -> None:
        """Mark headers as already initialized (when headers are set externally)."""
        self.headersSet = True
        self.connected = True
        self._log(f"Headers marked as initialized for session: {self.sessionId}")

    def sendProgress(self, message: str) -> None:
        """Send a progress message to the client."""
        self._sendEvent('progress', message)

    def sendResponse(self, data: dict) -> None:
        """Send a response (JSON data) to the client."""
        self._sendEvent('response', dumps(data))

    def sendError(self, message: str, code: int = 500) -> None:
        """Send an error to the client."""
        self._sendEvent('error', dumps({
            'error': True,
            'message': message,
            'code': code,
        }))

    def sendChunk(self, text: str) -> None:
        """Send a chunk of streaming text."""
        # DEBUG: Log EXACT data being sent to frontend (json_encode shows \n as \\n)
        error_log(f"📤 SENDING TO FRONTEND (len={len(text)}): " + dumps(text))
        self._sendEvent('chunk', text)

    def complete(self) -> None:
        """Signal that streaming is complete."""
        self._sendEvent('complete', dumps({'status': 'done'}))
        self.connected = False

    def sendCustomEvent(self, eventName: str, data: dict) -> None:
        """Send a custom event with arbitrary data."""
        self._sendEvent(eventName, dumps(data))

    def getSessionId(self) -> str:
        """Get the session ID."""
        return self.sessionId

    def isConnected(self) -> bool:
        """Check if the client is connected."""
        stream = self._get_stream()
        return self.connected and not (stream is not None and stream.aborted)

    def _sendEvent(self, event: str, data: str) -> None:
        """Send an SSE event."""
        if self.hubUrl:
            self._sendToHub(event, data)
        else:
            self._sendDirect(event, data)

    def _sendDirect(self, event: str, data: str) -> None:
        """Send event directly via the stream (for direct streaming)."""
        if not self.headersSet:
            self.headersSet = True
            self.connected = True
            self._log(f"SSE headers initialized for session: {self.sessionId}")

        stream = self._get_stream()
        if stream is not None:
            stream.send(event, data)

        self._log(f"Direct SSE sent - event: {event}")

    def _sendToHub(self, event: str, data: str) -> None:
        """Send event to an SSE hub server."""
        try:
            payload = {
                'session_id': self.sessionId,
                'event': event,
                'data': data,
            }
            r = httpx.post(self.hubUrl, json=payload, timeout=5)
            if r.status_code != 200:
                self._log(f"Hub send failed with HTTP {r.status_code}")
        except Exception as e:
            self._log("Hub send error: " + str(e))

    def _log(self, message: str) -> None:
        """Log a debug message."""
        if self.debug:
            error_log(f"[SSEHubClient] {message}")
